multihash/multihash.go

```diff
@@ -0,0 +1,48 @@
+package multihash
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"hash"
+)
+
+// Registry is a simple map which maps a multihash indicator number
+// to a standard golang Hash interface.
+//
+// Multihash indicator numbers are reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+// The keys used in this map must match those reservations.
+//
+// Hashers which are available in the golang stdlib are registered here automatically.
+//
+// Packages which want to register more hashing functions (and have a multihash number reserved!)
+// are encouraged to do so at package init time.
+// (Doing this at package init time ensures this map can be accessed without race conditions.)
+//
+// The linking/cid.DefaultLinkSystem will use this map to find hashers
+// to use when serializing data and computing links,
+// and when loading data from storage and verifying its integrity.
+//
+// This registry map is only used for default behaviors.
+// If you don't want to rely on it, you can always construct your own LinkSystem.
+// (For this reason, there's no special effort made to detect conflicting registrations in this map.
+// If more than one package registers for the same multicodec indicator, and
+// you somehow end up with both in your import tree, and yet care about which wins:
+// then just don't use this registry anymore: make a LinkSystem that does what you need.)
+// This should never be done to make behavior alterations
+// (hash functions are well standardized and so is the multihash indicator table),
+// but may be relevant if one is really itching to try out different hash implementations for performance reasons.
+var Registry = make(map[uint64]func() hash.Hash)
```

Comment:
The reality is there are a bunch of 3rd party hashes that end up in the current multihash registry in ipfs world and won't conform to hash.Hash

How do we expect to support that backwards compatibility?

Reply:
With all the terrible glue code I just wrote.  ┐_(ツ)_┌

See the new gadgets in `multihash/register/sha3` for handling Shake, for example.

(Like actually, review appreciated.  I just utterly invented a BlockSize which is nonsensical.  My theory is that it's harmless.)

I'll admit, I had kind of a discovery process today.  I didn't realize how many things in golang don't conform to `hash.Hash` directly.

At the same time: I think still more things conform to `hash.Hash` than conform to anything _else_, so I think sticking to that as the centroid is probably the best option available.

Reply:
again - isn't the reality that we need to take existing 3rd party codecs currently being registered with mulithash, and be able to keep them working after this transition? not "write a custom shim for each case" but be backwards compatible for them? If that grows to a week+ of work, it's less likely to make it into this refactor.

Reply:
Oh, "3rd party" in the sense of beyond-the-third-party-things-already-directly-depended-on-by-go-mulitcodec.  Mmm.  Okay, so, I think I had stated earlier that this was impossible, and I think I walk that back.  On further look, it seems possible; I just didn't see how at first, because it requires multiple steps, not just the so-called register function in that library.

But I still suspect, looking over the flavor of some of the issues in that repo, and how issues and PRs seem to focus on getting new things added to the library, that the number of third-party registrations made in the wild is asymptotically close to zero.

So I think at the moment: we can functionally disregard the possibility.  And if there are any applications out there which hit a speedbump here, they hit it when converting to go-ipld-prime... which is probably already in their roadmap as a speedbump, so there shouldn't be a category-change in speedbumpism, and so as a result I think everyone involved in such a hypothetical will probably survive.

Longer run: yeah, conversation is needed about how to re-reconcile this with go-multihash.  I think a bigger venue than this PR is going to be needed for that, though.